### ISSUE TITLE
`PreferredInterfaceType` false positive involving caller-dependent return type

### ISSUE THREAD

**Issue by Stephan202** [google/error-prone#3294]
Consider the following code:
```java
import com.google.common.collect.ImmutableMap;
import java.util.HashMap;
import java.util.Map;

final class Example {
  private final Map<String, Object> context = new HashMap<>();

  void setFoo(Map<String, String> map) {
    context.put("foo", map);
  }

  Map<String, String> getFoo() {
    return getContextOrDefault("foo", ImmutableMap.of());
  }

  @SuppressWarnings("unchecked")
  private <T> T getContextOrDefault(Object key, T defaultValue) {
    return !context.containsKey(key) ? defaultValue : (T) context.get(key);
  }
}
```

Compile the code as follows:
```sh
javac \
  -XDcompilePolicy=simple \
  -processorpath error_prone_core-2.14.0-with-dependencies.jar \
  -cp guava-31.1-jre.jar \
  '-Xplugin:ErrorProne -XepDisableAllChecks -Xep:PreferredInterfaceType:WARN' \
  Example.java
```

This yields the following warning:
```
Example.java:12: warning: [PreferredInterfaceType] Method return type should use the immutable type (such as ImmutableList) instead of the general collection interface type (such as List).
  Map<String, String> getFoo() {
     ^
    (see https://errorprone.info/bugpattern/PreferredInterfaceType)
  Did you mean 'ImmutableMap<String, String> getFoo() {'?
```

In this case changing `getFoo`'s return type can cause a `ClassCastException` at runtime, if `setFoo` is called with a non-`ImmutableMap` value prior to invoking `getFoo`.

The example code above is a simplification of an actual case we hit involving Project Reactor's [ContextView#getOrDefault](https://github.com/reactor/reactor-core/blob/ef67c7f6441e1dc33ffd160f4d20bdbdbf6067cd/reactor-core/src/main/java/reactor/util/context/ContextView.java#L75-L90) method.

NB: `PreferredInterfaceType` also triggers if one updates the example `getContextOrDefault` implementation to this arguably more "honest" alternative:
```diff
diff --git a/Example.java b/Example.java
index 4257330..26fc2f4 100644
--- a/Example.java
+++ b/Example.java
@@ -14,7 +14,7 @@ final class Example {
   }
 
   @SuppressWarnings("unchecked")
-  private <T> T getContextOrDefault(Object key, T defaultValue) {
-    return !context.containsKey(key) ? defaultValue : (T) context.get(key);
+  private <S, T extends S> S getContextOrDefault(Object key, T defaultValue) {
+    return !context.containsKey(key) ? defaultValue : (S) context.get(key);
   }
 }
```

Ideally `PreferredInterfaceType` is updated not to flag the case above. (I'm struggling to suggest clear heuristics, but wanted to file this ticket already in case others hit the same issue.)

**Comment by Stephan202** [comment]
Indeed `MixedMutabilityReturnType` doesn't trigger for our case. Changing our `getFoo()` implementation is something I considered (modifying the `setFoo` analog isn't really an option), but in this particular case we know that the retrieved data won't be modified, so an immutable default is safe (and slightly more efficient than creating a new mutable object for each invocation).

(But I have yet to find a better heuristic for the checker, so perhaps this issue should simply serve as a warning to future readers and be closed as "won't fix"/"can't fix" :thinking:.)

**Comment by graememorgan** [comment]
If you substitute the type variables, your generic method is:

```
  private ImmutableMap<String, String> getContextOrDefault(Object key, ImmutableMap<String, String> defaultValue) {
    return !context.containsKey(key) ? defaultValue : (ImmutableMap<String, String>) context.get(key);
  }
```

That _would_ throw a `ClassCastException` due to the cast to `ImmutableMap<String, String>`. Why isn't it? Well, generic type erasure. A cast to `(T)` is a cast to a type variable. The compiler can't in general know what it resolves to, so it gets elided.

I'd suggest your best remedy is to explicitly provide what you _want_ that cast to be:

```
  Map<String, String> getFoo() {
    return this.<Map<String, String>>getContextOrDefault("foo", ImmutableMap.of());
  }
```
